File: app/src/main/java/in/digistorm/aksharam/util/LanguageDetector.java
package in.digistorm.aksharam.util;

/*
 * Copyright (c) 2022 Alan M Varghese <alan@digistorm.in>
 *
 * This files is part of Aksharam, a script teaching app for Indic
 * languages.
 *
 * Aksharam is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Aksharam is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even teh implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import android.content.Context;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

public class LanguageDetector {
    private final String logTag = getClass().getSimpleName();
    private final LinkedHashMap<String, Language> languages;

    public LanguageDetector(Context context) {
        languages = LangDataReader.getAllLanguages(context);
        if(languages.isEmpty()) {
            Log.d(logTag, "Could not find any data files.");
        }
    }

    public String detectLanguage(String input, Context context) {
        Log.d(logTag, "Detecting language for " + input);
        Log.d(logTag, "languages loaded: " + languages.keySet());

        // now, we attempt to detect the input language
        String[] langs = languages.keySet().toArray(new String[0]);
        HashMap<String, Integer> score = new HashMap<>();
        for (char ch: input.toCharArray()) {
            for (Map.Entry<String, Language> entry: languages.entrySet()) {
                LetterDefinition letterDefinition = entry.getValue().getLetterDefinition(ch + "");
                if (letterDefinition != null) {
                    Log.d(logTag, " " + ch + " matches a character in the " + entry.getKey() + " set.");
                    if (score.containsKey(entry.getKey())) {
                        score.put(entry.getKey(), score.get(entry.getKey()) + 1);
                    }
                    else
                        score.put(entry.getKey(), 1);
                }
            }
        }

        String langDetected = null;
        int max_score = 0;
        for (Map.Entry<String, Integer> entry: score.entrySet()) {
            if (langDetected == null) {
                langDetected = entry.getKey();
                max_score = entry.getValue();
                continue;
            }
            if (entry.getValue() > max_score) {
                langDetected = entry.getKey();
                max_score = entry.getValue();
            }
        }
        Log.d(logTag, "Detected " + langDetected + " in input string with score " + max_score);

        return langDetected;
    }
}

File: app/src/main/kotlin/in/digistorm/aksharam/activities/main/language/LanguageDetector.kt
/*
 * Copyright (c) 2022 Alan M Varghese <alan@digistorm.in>
 *
 * This files is part of Aksharam, a script teaching app for Indic
 * languages.
 *
 * Aksharam is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Aksharam is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even teh implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package `in`.digistorm.aksharam.activities.main.language

import android.content.Context
import `in`.digistorm.aksharam.activities.main.helpers.upperCaseFirstLetter
import `in`.digistorm.aksharam.activities.main.util.logDebug
import java.util.HashMap
import java.util.LinkedHashMap

class LanguageDetector(context: Context?) {
    private val logTag = javaClass.simpleName
    private val languages: LinkedHashMap<String, Language>

    // Assumption: characters in a language are unique and each character exists only once
    // in all language definitions combined. That is we detect a language by its script. If two
    // languages use the same script, we cannot distinguish between them. Such a scenario is not
    // supported in this app, since we are primarily an app for script learning (not language learning).
    fun detectLanguage(input: String): String? {
        logDebug(logTag, "Detecting language for $input")
        logDebug(logTag, "languages loaded: " + languages.keys)

        // Now, we attempt to detect the input language
        val score = HashMap<String, Int>()
        outer@for (ch in input.toCharArray()) {
            for ((languageName, value) in languages) {
                val letterDefinition = value.getLetterDefinition(ch.toString())
                // If a definition exists for letter `ch` in language...
                if (letterDefinition != null) {
                    // If we are already tracking `languageName`, increment it. Else, initialise it to 1.
                    score[languageName] = score[languageName]?.inc() ?: 1
                    // We have found the letter, no point checking other languages.
                    // Proceed with next character.
                    continue@outer
                }
            }
            // We checked all languages and did not find this character.
            logDebug(logTag, "Character $ch in $input is unknown to us. Ignoring.")
        }
        var langDetected: String? = score.keys.firstOrNull()
        var maxScore = score[langDetected] ?: 0
        for ((key, value) in score) {
            if (value > maxScore) {
                langDetected = key
                maxScore = value
            }
        }
        logDebug(logTag, "Detected language $langDetected in string $input.")
        return langDetected?.upperCaseFirstLetter()
    }

    init {
        languages = getAllLanguages(context!!)
        if (languages.isEmpty()) {
            logDebug(logTag, "Could not find any data files.")
        }
    }
}
